### PR TITLE
Inventory item mousewheel movement

### PR DESCRIPTION
**Pull request by HybridDog**
Move your cursor above an item,
mousewheel down n times to take n items from the stack,
move your cursor above a free slot,
mousewheel up m times to put m items to the slot.

### CODE REVIEW

File: src/guiFormSpecMenu.cpp [hunk 1]
```diff
@@ -3311,24 +3311,30 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
...
+			button = 3 + (event.MouseInput.Wheel > 0); updown = 0; break;
+		case EMIE_LMOUSE_LEFT_UP:
+			button = 0; updown = 1; break;
+		case EMIE_RMOUSE_LEFT_UP:
+			button = 1; updown = 1; break;
+		case EMIE_MMOUSE_LEFT_UP:
+			button = 2; updown = 1; break;
+		case EMIE_MOUSE_MOVED:
+			updown = -1;
+		default:;
```
Comment (SmallJoker) [inline src/guiFormSpecMenu.cpp]: No default required

Comment (HybridDog) [inline src/guiFormSpecMenu.cpp]: l know, but it says a compiler warning if l omit it.

**Comment by HybridDog** [comment]
Description updated, there're 6 approvals.

File: src/guiFormSpecMenu.cpp (hunk 1 above)
Comment (SmallJoker) [inline src/guiFormSpecMenu.cpp]: Works fine here: https://github.com/minetest/minetest/blob/master/src/chat.cpp#L619
Perhaps try adding a `break;` to the line above?

Comment (HybridDog) [inline src/guiFormSpecMenu.cpp]: >src/guiFormSpecMenu.cpp: In member function ‘virtual bool GUIFormSpecMenu::OnEvent(const irr::SEvent&)’:
src/guiFormSpecMenu.cpp:3319:10: warning: enumeration value ‘EMIE_LMOUSE_DOUBLE_CLICK’ not handled in switch [-Wswitch]
   switch (event.MouseInput.Event) {
          ^
src/guiFormSpecMenu.cpp:3319:10: warning: enumeration value ‘EMIE_RMOUSE_DOUBLE_CLICK’ not handled in switch [-Wswitch]
src/guiFormSpecMenu.cpp:3319:10: warning: enumeration value ‘EMIE_MMOUSE_DOUBLE_CLICK’ not handled in switch [-Wswitch]
src/guiFormSpecMenu.cpp:3319:10: warning: enumeration value ‘EMIE_LMOUSE_TRIPLE_CLICK’ not handled in switch [-Wswitch]
src/guiFormSpecMenu.cpp:3319:10: warning: enumeration value ‘EMIE_RMOUSE_TRIPLE_CLICK’ not handled in switch [-Wswitch]
src/guiFormSpecMenu.cpp:3319:10: warning: enumeration value ‘EMIE_MMOUSE_TRIPLE_CLICK’ not handled in switch [-Wswitch]
src/guiFormSpecMenu.cpp:3319:10: warning: enumeration value ‘EMIE_COUNT’ not handled in switch [-Wswitch]

a break isn't requisite at the end

**Comment by Wuzzy2** [comment]
While I approve of the idea very much:

Sigh. Another hard-coded key which users can't reconfigure. :-(

File: src/guiFormSpecMenu.cpp [hunk 2]
```diff
@@ -3311,24 +3311,30 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
...
+		case EMIE_MOUSE_WHEEL:
+			button = 3 + (event.MouseInput.Wheel > 0); updown = 0; break;
+		case EMIE_LMOUSE_LEFT_UP:
+			button = 0; updown = 1; break;
+		case EMIE_RMOUSE_LEFT_UP:
+			button = 1; updown = 1; break;
+		case EMIE_MMOUSE_LEFT_UP:
+			button = 2; updown = 1; break;
+		case EMIE_MOUSE_MOVED:
+			updown = -1;
```
Comment (sfan5) [inline src/guiFormSpecMenu.cpp]: missing a `break`

(no, i dont care that it works fine without)

File: src/guiFormSpecMenu.cpp [hunk 3]
```diff
@@ -3361,73 +3367,80 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
...
-					u32 count;
+					u32 count = 0;
 					if (button == 1)  // right
 						count = (s_count + 1) / 2;
 					else if (button == 2)  // middle
 						count = MYMIN(s_count, 10);
-					else  // left
+					else if (button == 3)  // wheeldown
+						count = 1;
+					else if (!button)  // left
```
Comment (sfan5) [inline src/guiFormSpecMenu.cpp]: why not keep this as `else  // left`?

File: src/guiFormSpecMenu.cpp [hunk 4]
```diff
@@ -3361,73 +3367,80 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
...
 
 				if (s.isValid()) {
 					// Clicked a slot: move
-					if (button == 1)  // right
+					if (button == 1 || button == 4)  // right or wheelup
 						move_amount = 1;
 					else if (button == 2)  // middle
 						move_amount = MYMIN(m_selected_amount, 10);
-					else  // left
+					else if (!button)  // left
```
Comment (sfan5) [inline src/guiFormSpecMenu.cpp]: again why change this?

File: src/guiFormSpecMenu.cpp [hunk 5]
```diff
@@ -3361,73 +3367,80 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
...
 				}
 				else if (!getAbsoluteClippingRect().isPointInside(m_pointer)) {
 					// Clicked outside of the window: drop
-					if (button == 1)  // right
+					if (button == 1 || button == 4)  // right or wheelup
 						drop_amount = 1;
 					else if (button == 2)  // middle
 						drop_amount = MYMIN(m_selected_amount, 10);
-					else  // left
+					else if (!button)  // left
```
Comment (sfan5) [inline src/guiFormSpecMenu.cpp]: same

File: src/guiFormSpecMenu.cpp [hunk 6]
```diff
@@ -3361,73 +3367,80 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
...
-				}
-			} else if (m_selected_item != NULL && m_selected_dragging &&
+			if (m_selected_dragging && m_selected_item != NULL) {
+				if (s.isValid()) {
+					if (!identical) {
+						// Dragged to different slot: move all selected
+						move_amount = m_selected_amount;
+					}
+				} else if (
 					!(getAbsoluteClippingRect().isPointInside(m_pointer))) {
```
Comment (sfan5) [inline src/guiFormSpecMenu.cpp]: put this on the previous line

File: src/guiFormSpecMenu.cpp (hunk 4 above)
Comment (HybridDog) [inline src/guiFormSpecMenu.cpp]: @sfan5 button ∈ ℤ₃ \ {1, 2} ≢ button ∈ ℤ₅ \ {1, 2, 4}

Comment (HybridDog) [inline src/guiFormSpecMenu.cpp]: the wheeldown needs to be ignored

**Comment by HybridDog** [comment]
Can this be merged now?

**Comment by HybridDog** [comment]
the king can hear you all

**Comment by HybridDog** [comment]
please feedback

File: src/guiFormSpecMenu.cpp [hunk 7]
```diff
@@ -3311,24 +3311,30 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
...
-			{ button = 0; updown = 1; }
-		else if (event.MouseInput.Event == EMIE_RMOUSE_LEFT_UP)
-			{ button = 1; updown = 1; }
-		else if (event.MouseInput.Event == EMIE_MMOUSE_LEFT_UP)
-			{ button = 2; updown = 1; }
-		else if (event.MouseInput.Event == EMIE_MOUSE_MOVED)
-			{ updown = -1;}
+		switch (event.MouseInput.Event) {
+		case EMIE_LMOUSE_PRESSED_DOWN:
+			button = 0; updown = 0; break;
```
Comment (nerzhul) [inline src/guiFormSpecMenu.cpp]: break on the next line

Comment (HybridDog) [reply src/guiFormSpecMenu.cpp]: done

**Comment by HybridDog** [comment]
‫@nerzhul‬ did l fix everything?

**Comment by HybridDog** [comment]
How can I get rid of the Changes requested?

**Comment by HybridDog** [comment]
@SmallJoker rebased, can I change some github setting to become notified when labels are added or removed?

**Comment by SmallJoker** [comment]
@HybridDog I'm not aware of any setting that would make this possible. Checking the pull requests sometimes on https://github.com/pulls is the most efficient method right now.

**Comment by HybridDog** [comment]
Please tell me if you have any objections about merging this.

**Comment by Fixer-007** [comment]
Groundbreaking, but other improvements needed as well!

**Pull request merged**